tuplex/codegen/src/IteratorContextProxy.cc

```diff
@@ -0,0 +1,494 @@
+            FunctionType *ft = llvm::FunctionType::get(llvm::Type::getInt1Ty(_env->getContext()),
+                                                       {llvm::PointerType::get(iteratorContextType, 0)}, false);
+            auto *nextFunc = _env->getModule()->getOrInsertFunction(funcName, ft).getCallee();
+            auto exhausted = builder.CreateCall(nextFunc, iterator);
+            return exhausted;
+        }
+
+        SerializableValue IteratorContextProxy::getIteratorNextElement(llvm::IRBuilder<> &builder,
+                                                                   const python::Type &yieldType,
+                                                                   llvm::Value *iterator,
+                                                                   IteratorInfo *iteratorInfo) {
+            using namespace llvm;
+
+            llvm::Type *iteratorContextType = iterator->getType()->getPointerElementType();
+            std::string funcName;
+            auto iteratorName = iteratorInfo->iteratorName;
+
+            if(iteratorName == "zip") {
+                return getZipNextElement(builder, yieldType, iterator, iteratorInfo);
+            }
+
+            if(iteratorName == "enumerate") {
+                return getEnumerateNextElement(builder, yieldType, iterator, iteratorInfo);
+            }
+
+            assert(iteratorName == "iter");
+            auto iterablesType = iteratorInfo->argsType;
+            auto argsIteratorInfo = iteratorInfo->argsIteratorInfo;
+            if(iterablesType.isIteratorType()) {
+                // iter() call on an iterator, ignore the outer iter and call again
+                assert(argsIteratorInfo.front());
+                return getIteratorNextElement(builder, yieldType, iterator, argsIteratorInfo.front());
+            }
+
+            // get current element value and size of current value
+            llvm::Value *retVal = nullptr, *retSize = nullptr;
+            auto indexPtr = builder.CreateGEP(iteratorContextType, iterator, {_env->i32Const(0), _env->i32Const(1)});
+            auto index = builder.CreateLoad(indexPtr);
+            auto iterableAllocPtr = builder.CreateGEP(iteratorContextType, iterator, {_env->i32Const(0), _env->i32Const(2)});
+            auto iterableAlloc = builder.CreateLoad(iterableAllocPtr);
+            if(iterablesType.isListType()) {
+                auto valArrayPtr = builder.CreateGEP(_env->getListType(iterablesType), iterableAlloc, {_env->i32Const(0), _env->i32Const(2)});
+                auto valArray = builder.CreateLoad(valArrayPtr);
+                auto currValPtr = builder.CreateGEP(valArray, index);
+                retVal = builder.CreateLoad(currValPtr);
+                if(yieldType == python::Type::I64 || yieldType == python::Type::F64) {
+                    retSize = _env->i64Const(8);
+                } else if(yieldType == python::Type::BOOLEAN) {
+                    retSize = _env->i64Const(1);
```

Comment:
I think technically we still use 8 bytes for a boolean. But the size field is never used.

Reply:
Revised.

Reply:
just as a side note, the list internal representation uses 1 byte for booleans (shouldn't be relevant here, but its a bit confusing because we haven't changed the regular representation yet).